### PR TITLE
feat: update time releaser option to release only inactive tasks that never had a heartbeat

### PR DESCRIPTION
**Pull request by nitzanmorr**
Added get heartbeat route to update time releaser's functionality. Now Task liberator will only release tasks that never had a heartbeat, so it won't take active tasks back to pending. This change should make the task liberator (updateTimeReleaser) liberate 'lost' tasks when they become in-progress while no worker is actually working on that task.

| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✔                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✔                                                                        |
| Chore            | ✖                                                                       |

### CODE REVIEW

File: src/updateTime/updateTimeReleaser.ts [hunk 1]
```diff
@@ -28,7 +31,17 @@ export class UpdateTimeReleaser {
...
     this.logger.info('starting update time releaser.');
 
-    const deadTasks = await this.tasksClient.getInactiveTasks();
+    const inactiveTasks = await this.tasksClient.getInactiveTasks();
+    const deadTasks: string[] = [];
+    for (const task of inactiveTasks) {
+      try {
+        await this.heartbeatClient.getHeartbeat(task);
+      } catch (error) {
+        if (error instanceof NotFoundError) {
```
Comment (CL-SHLOMIKONCHA) [inline src/updateTime/updateTimeReleaser.ts:40]: please add debug log for the task that found as dead before push it

File: src/updateTime/updateTimeReleaser.ts [hunk 2]
```diff
@@ -5,18 +5,23 @@ import { TasksClient } from '../clients/tasksClient';
...
   public constructor(
     @inject(Services.CONFIG) config: IConfig,
     @inject(Services.LOGGER) private readonly logger: Logger,
     @inject(Services.TRACER) private readonly tracer: Tracer,
-    private readonly tasksClient: TasksClient
+    private readonly tasksClient: TasksClient,
+    private readonly heartbeatClient: HeartbeatClient
   ) {
     this.enabled = toBoolean(config.get('updateTime.enabled'));
+    this.checkHeartbeatEnabled = toBoolean(config.get('updateTime.checkHeartbeat'));
```
Comment (CL-SHLOMIKONCHA) [inline src/updateTime/updateTimeReleaser.ts:24]: `toBoolean()` function seems unnecessary here

Comment (CL-SHLOMIKONCHA) [reply src/updateTime/updateTimeReleaser.ts:24]: image

**Review by CL-SHLOMIKONCHA — CHANGES_REQUESTED** [review]
Very good job!

File: src/updateTime/updateTimeReleaser.ts (hunk 2 above)
Comment (CL-SHLOMIKONCHA) [inline src/updateTime/updateTimeReleaser.ts:24]:
```suggestion
    this.checkHeartbeatEnabled = config.get<boolean>('updateTime.checkHeartbeat');
```

Comment (nitzanmorr) [reply src/updateTime/updateTimeReleaser.ts]: fixed

Comment (nitzanmorr) [reply src/updateTime/updateTimeReleaser.ts]: fixed and deleted common/utilities/typeConvertors.ts that contained toBoolean function that is no longer in use

**Review by CL-SHLOMIKONCHA — APPROVED** [review]

**Review by netanelC — CHANGES_REQUESTED** [review]
Good PR! I am impressed with the config mock that you added!
You added very important changes, but separate the changes so only one context (feature/issue) will be in a PR so it will be easier to relate the changes to the PR purpose.

File: src/updateTime/updateTimeReleaser.ts [hunk 3]
```diff
@@ -28,7 +32,22 @@ export class UpdateTimeReleaser {
...
-    const deadTasks = await this.tasksClient.getInactiveTasks();
+    const inactiveTasks = await this.tasksClient.getInactiveTasks();
+    let deadTasks: string[] = [];
+    if (this.checkHeartbeatEnabled) {
+      for (const task of inactiveTasks) {
+        try {
+          await this.heartbeatClient.getHeartbeat(task);
+        } catch (error) {
+          if (error instanceof NotFoundError) {
+            this.logger.debug(`found dead job that never had a heartbeat: ${task}`);
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/updateTime/updateTimeReleaser.ts:43]: fixed

File: tests/mocks/config.ts [hunk 4]
```diff
@@ -1,11 +1,70 @@
...
+
+const setConfigValues = (values: Record<string, unknown>): void => {
+  getMock.mockImplementation((key: string) => {
+    const value = get(values, key) ?? config.get(key);
+    return value;
+  });
+  hasMock.mockImplementation((key: string) => has(values, key) || config.has(key));
+};
+
+const registerDefaultConfig = (): void => {
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply tests/mocks/config.ts:40]: removed this function

File: tests/unit/heartbeat/heartbeatReleaser.spec.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import jsLogger from '@map-colonies/js-logger';
 import { HeartbeatReleaser } from '../../../src/heartbeat/heartbeatReleaser';
 import { tracerMock, initTrace } from '../../mocks/openTelemetry/tracer';
-import { configMock, getMock } from '../../mocks/config';
+import { configMock, getMock, setValue, init as initconfig, clear as clearConfig } from '../../mocks/config';
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply tests/unit/heartbeat/heartbeatReleaser.spec.ts:4]: fixed

File: tests/unit/updateTime/updateTimeReleaser.spec.ts [hunk 6]
```diff
@@ -1,67 +1,121 @@
...
       expect(getMock).toHaveBeenCalledWith('updateTime.enabled');
-      expect(getMock).toHaveBeenCalledTimes(1);
+      expect(getMock).toHaveBeenCalledWith('updateTime.checkHeartbeat');
+      expect(getMock).toHaveBeenCalledTimes(2);
       expect(tasksInactiveTasksMock).toHaveBeenCalledTimes(1);
       expect(tasksReleaseTasksMock).not.toHaveBeenCalled();
     });
 
-    it('release dead tasks', async function () {
+    it('release dead task when heartbeat check is disabled and task is inactive but had a heartbeat', async function () {
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply tests/unit/updateTime/updateTimeReleaser.spec.ts:56]: added another describe block for when heartbeat check is disabled

File: package.json [hunk 7]
```diff
@@ -48,7 +48,8 @@
     "http-status-codes": "^2.2.0",
     "lodash": "^4.17.21",
     "reflect-metadata": "^0.1.13",
-    "tsyringe": "^4.7.0"
+    "tsyringe": "^4.7.0",
+    "uuid": "^10.0.0"
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply package.json:52]: migrated to faker

File: tests/mocks/config.ts [hunk 8]
```diff
@@ -1,11 +1,32 @@
+import config from 'config';
+import { get, has } from 'lodash';
```
Comment (CL-SHLOMIKONCHA) [inline tests/mocks/config.ts:2]: defined but not in use

File: tests/mocks/config.ts [hunk 9]
```diff
@@ -1,11 +1,70 @@
...
+    expirationStatus: {
+      enabled: true,
+    },
+    httpRetry: {
+      attempts: 3,
+      delay: 'exponential',
+      shouldResetTimeout: true,
+    },
+  };
+  setConfigValues(config);
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply tests/mocks/config.ts]: @nitzanmorr removed.

File: tests/mocks/config.ts [hunk 10]
```diff
@@ -1,11 +1,70 @@
...
+};
+
+export { getMock, hasMock, configMock, setValue, clear, init, setConfigValues, registerDefaultConfig };
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply tests/mocks/config.ts]: @nitzanmorr removed

File: tests/unit/taskExpiration/expirationStatusUpdater.spec.ts [hunk 11]
```diff
@@ -1,24 +1,27 @@
 import jsLogger from '@map-colonies/js-logger';
 import { ExpirationStatusUpdater } from '../../../src/taskExpiration/expirartionStatusUpdater';
 import { tracerMock, initTrace } from '../../mocks/openTelemetry/tracer';
-import { configMock, getMock } from '../../mocks/config';
+import { configMock, getMock, setValue, init as initconfig, clear as clearConfig } from '../../mocks/config';
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply tests/unit/taskExpiration/expirationStatusUpdater.spec.ts]: @nitzanmorr missed here

**Review by CL-SHLOMIKONCHA — CHANGES_REQUESTED** [review]

File: tests/mocks/config.ts (hunk 8 above)
Comment (nitzanmorr) [reply tests/mocks/config.ts:2]: fixed

**Pull request merged**